File: NeetCode/10-Heap-Priority-Queue/LastStoneWeight.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
class Solution {  
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int, vector<int>> heap; 
        for(int stone : stones){
            heap.push(stone);
        }
        while(heap.size() > 1){
            int heavy = heap.top();
            heap.pop();
            int secondHeavy = heap.top();
            heap.pop();
            if(heavy != secondHeavy){
                heap.push(heavy-secondHeavy);
            } 
        }
        if(heap.empty()){
            return 0;
        }
        return heap.top();
    }
};

File: NeetCode/10-Heap-Priority-Queue/LastStoneWeight.java
import java.util.*;
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> heap = new PriorityQueue<>(Collections.reverseOrder());
        for(int stone : stones){
            heap.offer(stone);
        }
        while(heap.size() > 1){
            int heavy = heap.poll();
            int secondHeavy = heap.poll();
            if(heavy != secondHeavy){
                heap.offer(heavy-secondHeavy);
            } 
        }
        if(heap.isEmpty()){
            return 0;
        }
        return heap.poll();
    }
}